mm2src/coins/eth.rs

```diff
@@ -6179,3 +6190,755 @@ async fn get_eth_gas_details(
+    }
+}
+
+#[async_trait]
+impl MakerNftSwapOpsV2 for EthCoin {
+    async fn send_nft_maker_payment_v2(
+        &self,
+        args: SendNftMakerPaymentArgs<'_, Self>,
+    ) -> Result<Self::Tx, TransactionErr> {
+        self.send_nft_maker_payment_v2_impl(args).await
+    }
+
+    async fn validate_nft_maker_payment_v2(
+        &self,
+        args: ValidateNftMakerPaymentArgs<'_, Self>,
+    ) -> ValidatePaymentResult<()> {
+        self.validate_nft_maker_payment_v2_impl(args).await
+    }
+
+    async fn spend_nft_maker_payment_v2(
+        &self,
+        args: SpendNftMakerPaymentArgs<'_, Self>,
+    ) -> Result<Self::Tx, TransactionErr> {
+        self.spend_nft_maker_payment_v2_impl(args).await
+    }
+
+    async fn refund_nft_maker_payment_v2_timelock(
+        &self,
+        args: RefundPaymentArgs<'_>,
+    ) -> Result<Self::Tx, TransactionErr> {
+        self.refund_nft_maker_payment_v2_timelock_impl(args).await
+    }
+
+    async fn refund_nft_maker_payment_v2_secret(
+        &self,
+        _args: RefundMakerPaymentArgs<'_, Self>,
+    ) -> Result<Self::Tx, TransactionErr> {
+        todo!()
+    }
+}
+
+impl EthCoin {
+    async fn send_nft_maker_payment_v2_impl(
+        &self,
+        args: SendNftMakerPaymentArgs<'_, Self>,
+    ) -> Result<SignedEthTx, TransactionErr> {
+        let contract_type = try_tx_s!(self.parse_contract_type(args.contract_type));
+        try_tx_s!(self.validate_payment_args(
+            args.taker_secret_hash,
+            args.maker_secret_hash,
+            &args.amount,
+            &contract_type,
+        ));
+
+        let taker_address = try_tx_s!(addr_from_raw_pubkey(args.taker_pub));
+        let token_address = try_tx_s!(self.parse_token_contract_address(args.token_address));
+        let swap_contract_address = try_tx_s!(self.parse_token_contract_address(args.swap_contract_address));
+        let time_lock_u32 = try_tx_s!(args.time_lock.try_into());
+        let token_id_u256 = U256::from(args.token_id);
+        let htlc_data = self.prepare_htlc_data(&args, taker_address, token_address, time_lock_u32);
+
+        match &self.coin_type {
+            EthCoinType::Nft { .. } => match contract_type {
+                ContractType::Erc1155 => {
+                    let function = try_tx_s!(ERC1155_CONTRACT.function("safeTransferFrom"));
+                    let amount_u256 = try_tx_s!(U256::from_dec_str(&args.amount.to_string()));
+                    let data = try_tx_s!(function.encode_input(&[
+                        Token::Address(self.my_address),
+                        Token::Address(swap_contract_address),
+                        Token::Uint(token_id_u256),
+                        Token::Uint(amount_u256),
+                        Token::Bytes(htlc_data),
+                    ]));
+                    self.sign_and_send_transaction(0.into(), Action::Call(token_address), data, U256::from(ETH_GAS))
+                        .compat()
+                        .await
```

Comment:
This call can be outside the match statement, then the match statement can be converted to a function called `prepare_maker_payment_v2_data` or any such name.

Reply:
Agree, I reduced code duplication in send, spend funcs here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2084/commits/6edcdb9196b0839f7926fa5518926bf852d6bcf4
and in validation impl here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2084/commits/5978908c5b69a3d35bcff71f0afda1be3fe7c3a0